modelforge/potential/aimnet2.py

```diff
@@ -168,54 +169,445 @@ def compute_properties(
+
+        return {
+            "per_atom_energy": E_i,
+            "per_atom_scalar_representation": atomic_embedding,
+            "atomic_subsystem_indices": data.atomic_subsystem_indices,
+        }
+
+
+class BaseMessageModule(nn.Module):
+    def __init__(self, number_of_per_atom_features: int):
+        """
+        Initialize the BaseMessageModule.
+
+        Parameters
+        ----------
+        number_of_per_atom_features : int
+            The number of features per atom.
+        """
+        super().__init__()
+        self.number_of_per_atom_features = number_of_per_atom_features
+
+        # Separate linear layers for embeddings and charges
+        self.linear_transform_embeddings = nn.Linear(
+            number_of_per_atom_features, number_of_per_atom_features
+        )
+        self.linear_transform_charges = nn.Linear(
+            1, number_of_per_atom_features
+        )  # For partial charges
+
+    def calculate_contributions(
+        self,
+        per_atom_feature_tensor: torch.Tensor,
+        pairlist: torch.Tensor,
+        f_ij_cutoff: torch.Tensor,
+        r_ij: torch.Tensor,
+        linear_transform: nn.Module,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Calculate the radial and vector contributions for the given features.
+
+        Parameters
+        ----------
+        per_atom_feature_tensor : torch.Tensor
+            The feature tensor (either atomic embeddings or repeated partial charges).
+        pairlist : torch.Tensor
+            The list of atom pairs.
+        f_ij_cutoff : torch.Tensor
+            The cutoff function applied to the radial symmetry functions.
+        r_ij : torch.Tensor
+            The displacement vectors between atom pairs.
+        linear_transform : nn.Module
+            The linear transformation to apply to the features.
+
+        Returns
+        -------
+        Tuple[torch.Tensor, torch.Tensor]
+            Radial and vector contributions.
+        """
+
+        idx_j = pairlist[1]
+
+        # Calculate the unit vector u_ij
+        r_ij_norm = torch.norm(r_ij, dim=1, keepdim=True)  # Shape: (num_atom_pairs, 1)
+        u_ij = r_ij / r_ij_norm  # Shape: (num_atom_pairs, 3)
+
+        # Step 1: Radial Contributions Calculation (Equation 4)
+        proto_v_r_a = (
+            f_ij_cutoff * per_atom_feature_tensor[idx_j]
+        )  # Shape: (num_atom_pairs, nr_of_features)
+
+        # Initialize tensor to accumulate radial contributions for each atom
+        radial_contributions = torch.zeros(
+            (per_atom_feature_tensor.shape[0], self.number_of_per_atom_features),
+            device=per_atom_feature_tensor.device,
+            dtype=per_atom_feature_tensor.dtype,
+        )  # Shape: (num_of_atoms, nr_of_features)
+
+        # Accumulate the radial contributions using index_add_
+        radial_contributions.index_add_(0, idx_j, proto_v_r_a)
+
+        # Step 2: Vector Contributions Calculation (Equation 5)
```

Comment:
This was the step we aren't completely certain about right? If so, a TODO comment would be a good visual indicator.

Reply:
will do!